File: nodejs/day10/index1.js
const { readFile, parseInt } = require("../lib");

const a = Array.from({ length: 256 }, (_, i) => i);
const L = a.length;

const ls = readFile("./input")
  .split(",")
  .map(parseInt);

let i = 0;
let skip = 0;

const read = l => {
  let max = i + l;
  const r = a.slice(i, max);
  if (max < L) {
    return r;
  }
  max -= L;
  return r.concat(a.slice(0, max));
};

const write = b => {
  let k = i;
  for (const el of b) {
    a[k] = el;
    k++;
    if (k >= L) {
      k -= L;
    }
  }
};

const revert = l => {
  write(read(l).reverse());
};

for (const l of ls) {
  revert(l);
  i = (i + l + skip) % L;
  skip++;
}

console.log(a[0] * a[1]);

File: clojure/day10/index1.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))

(defn split [p s] (s/split s p))

(defn cnct [& as] (into [] (apply concat as)))

(def ls (->> "./input" slurp s/trim-newline (split #",") (map parse-int)))

(def L 256)

(def a (into [] (range L)))

(defn read-slice [a l i] (let [
  L (count a)
  m (+ i l)
  m' (min m L)
  r (subvec a i m')
](if (<= m L) r
  (cnct r (subvec a 0 (- m L))))))

(defn write-slice [a b i] (let [
  L (count a)
] (
  loop [
    a a
    i i
    rem-b b
  ]
    (if (== 0 (count rem-b)) a
      (let [
        el (first rem-b)
        rem-b' (rest rem-b)
        i' (mod i L)
        a' (assoc a i' el)
      ]
        (recur a' (inc i') rem-b'))
  ))))

(defn transform [a l i]
  (let [
    r (read-slice a l i)
    r' (reverse r)
  ] (write-slice a r' i)))

(defn run-op [a i skip l]
  (let [L (count a)]
    [(transform a l i) (mod (+ i l skip) L) (inc skip)]))

(def a' (loop [
  a a
  i 0
  skip 0
  rem-ls ls
] (if (== 0 (count rem-ls)) a
  (let [
    l (first rem-ls)
    rem-ls' (rest rem-ls)
    [a' i' skip'] (run-op a i skip l)
  ] (recur a' i' skip' rem-ls')))))

(def res (* (first a') (second a')))

(println res)
